### ISSUE TITLE
Conditional Mapping

### ISSUE THREAD

**Issue by RoyC-Retail-Assist** [linq2db/linq2db#1219]
I have dates stored in the database as DECIMAL(8,0) yyyyMMdd format and times as DECIMAL(6,0) hhmmss.

I'd like to be able to use these as DateTime however there are other DateTime values that are actually stored as DateTime on the database..
```cs
public class Test
{
	public int Id { get; set; }

	[Column(DataType = LinqToDB.DataType.Decimal, Length = 8, Precision = 0)]
	public DateTime Date { get; set; }
}
```

For reading data I can set a converter.
```cs
MappingSchema.Default.SetConverter<decimal, DateTime>(val =>
{
	if (val >= 10000101 && val < 99991231)
	{
		if (DateTime.TryParseExact(val.ToString(), "yyyyMMdd", null, System.Globalization.DateTimeStyles.None, out var retval))
			return retval;
	}

	if (val > 0 && val < 240000)
	{
		if (DateTime.TryParseExact(val.ToString(), "hhmmss", null, System.Globalization.DateTimeStyles.None, out var retval))
			return retval;
	}

	return DateTime.MinValue;
});
```

I can set a converter for writing data but this will apply to any DateTime > DataParameter:
```cs
MappingSchema.Default.SetConverter<DateTime, DataParameter>((obj) =>
{
	return new DataParameter { DataType = DataType.Decimal, Value = GetDecimalDate(obj) };
});
```

Is there any way of having a converter that is conditional upon _something_ so that any users of the class don't have to do any special to write the data? Alternatively, is there a better way of achieving this?

Hopefully these are dumb questions 😇 

### Environment details
linq2db version: 2.1.0
Database Server: Any
Database Provider: Any
Operating system: Windows 10
Framework version: .NET Framework 4.6.2

**Comment by Franki1986** [comment]
Why don't add a custom attribute so that you can evaluate when decimal should be used?

**Comment by RoyC-Retail-Assist** [comment]
The Column attribute has enough details to use but the Converter doesn't, as far as I can see, have any reference to the target column so the attribute cannot be accessed.  I'm guessing that there might be something further up the expression tree that can be utilised but haven't been able to find it yet hence my question.

**Comment by sdanyliv** [comment]
@RoyC-Retail-Assist, how do you plan to use such classes with data fields? I assume only on client side.
Maybe better to extend your T4 template and generate additional readonly properties that reads data in appropriate way.

**Comment by RoyC-Retail-Assist** [comment]
Ideally we would be able to read and write the columns but also to be able to use them as part of a Where, GroupBy and/or OrderBy on the server side.  We have been using calculated columns for this but these do not allow the server side selections.

We don't use T4 templates to generate our data classes but can code anything that is sensible to achieve this.

The DateTime value is passed as a parameter and I was hoping that when the parameter is initially created then the Column attribute DataType could be checked to then add the appropriate conversion for DateTime -> Decimal.


Although this is driver agnostic I also have a specific requirement in the DB2 iSeries provider to be able to map DateTime to LinqToDb.DataType.Date OR LinqToDb.DataType.Date dependant upon the ColumnAttribute but by the time the code in the provider is hit the underlying column is not available.

**Comment by sdanyliv** [comment]
Will think. BTW, update ProviderDetector :)

**Comment by RoyC-Retail-Assist** [comment]
I working on it now

**Comment by sdanyliv** [comment]
@RoyC-Retail-Assist, how about such syntax? For sure it is not supported yet. Just thinking:
```cs
MappingSchema.Default.SetConverter<DateTime, DataParameter>(DataType.Decimal, (obj) =>
{
    return new DataParameter { DataType = DataType.Decimal, Value = GetDecimalDate(obj) };
}); 
```

**Comment by RoyC-Retail-Assist** [comment]
That syntax would be good :+1:

I'm assuming that this would be matching the DataType parameter to the DataType specified by the Column Attribute to determine that the converter is to be used.  

If that is the case then it might also be worthwhile adding a MappingType string parameter to the ColumnAttribute and having another SetConverter overload that uses that.  This would allow completely custom mapping based upon the Column attribute.

In this particular use case we could use MappingType = "DecimalTime" to convert the DateTime value into a decimal hhmmss and "DecimalDate" to yyyymmdd.

**Comment by MaceWindu** [comment]
Closed by #1263
